src/PhpDoc/TypeNodeResolver.php

```diff
@@ -1081,6 +1073,69 @@ private function resolveArrayShapeNode(ArrayShapeNode $typeNode, NameScope $name
 		return $arrayType;
 	}
 
+	private function resolveArrayShapeOffsetType(ArrayShapeItemNode $itemNode, NameScope $nameScope): ?Type
+	{
+		if ($itemNode->keyName instanceof ConstExprIntegerNode) {
+			return new ConstantIntegerType((int) $itemNode->keyName->value);
+		} elseif ($itemNode->keyName instanceof IdentifierTypeNode) {
+			return new ConstantStringType($itemNode->keyName->name);
+		} elseif ($itemNode->keyName instanceof ConstExprStringNode) {
+			return new ConstantStringType($itemNode->keyName->value);
+		} elseif ($itemNode->keyName instanceof ConstFetchNode) {
+			$constExpr = $itemNode->keyName;
+			if ($constExpr->className === '') {
+				throw new ShouldNotHappenException(); // global constant should get parsed as class name in IdentifierTypeNode
+			}
+
+			if ($nameScope->getClassName() !== null) {
+				switch (strtolower($constExpr->className)) {
+					case 'static':
+					case 'self':
+						$className = $nameScope->getClassName();
+						break;
+
+					case 'parent':
+						if ($this->getReflectionProvider()->hasClass($nameScope->getClassName())) {
+							$classReflection = $this->getReflectionProvider()->getClass($nameScope->getClassName());
+							if ($classReflection->getParentClass() === null) {
+								return new ErrorType();
+
+							}
+
+							$className = $classReflection->getParentClass()->getName();
+						}
+						break;
+				}
+			}
+
+			if (!isset($className)) {
+				$className = $nameScope->resolveStringName($constExpr->className);
```

Comment:
I think this is a bit undertested - all these branches - self/static/parent/normal name should have a test in `tests/PHPStan/Analyser/nsrt/bug-6989.php`.

Reply:
I added a test for all the possible case I think.